### PR TITLE
contract standards: add ext_fungible_token

### PR DESCRIPTION
**Pull request by robert-zaremba**
When implementing contracts which operate with fungible tokens, it's useful to use `#[ext_contract(ext_fungible_token)]` macro.

This PR adds a canonical implementation of that wrapper to near-contract-standards.

### CODE REVIEW

**Comment by evgenykuzyakov** [comment]
I wonder if it adds all methods to contract storage even if you're using only one method. But because they are static, it probably doesn't.

File: near-contract-standards/src/fungible_token/core_impl.rs [hunk 1]
```diff
@@ -33,6 +33,25 @@ pub trait FungibleTokenReceiver {
     ) -> PromiseOrValue<U128>;
 }
 
+#[ext_contract(ext_fungible_token)]
+pub trait FungibleTokenContract {
+    fn ft_transfer(&mut self, receiver_id: AccountId, amount: U128, memo: Option<String>);
```
Comment (evgenykuzyakov) [inline near-contract-standards/src/fungible_token/core_impl.rs:38]: Should we use `ValidAccountId` for consistency? I wish we didn't need to do conversion from `AccountId` to `ValidAccountId`, so you don't have to every operate on just `AccountId` unless you do sting operations.
```suggestion
    fn ft_transfer(&mut self, receiver_id: ValidAccountId, amount: U128, memo: Option<String>);
```

Comment (robert-zaremba) [reply near-contract-standards/src/fungible_token/core_impl.rs:38]: yeah, in smart contract, we usually are after validation, so converting back to `ValidAccountId` is usually just an unnecessary step. If you think it's better to have it with a type system, then feel free to commit the suggestion. I don't have a strong opinion.

**Comment by robert-zaremba** [comment]
> I wonder if it adds all methods to contract storage even if you're using only one method. But because they are static, it probably doesn't.

You mean to the binary? Everything what is static and not used  (structures, static functions, const functions) should be removed in the compile time.

File: near-contract-standards/src/fungible_token/core_impl.rs (hunk 1 above)
Comment (evgenykuzyakov) [reply near-contract-standards/src/fungible_token/core_impl.rs:38]: Yeah, let's use `AccountId` then in 2 places below

File: near-contract-standards/src/fungible_token/core_impl.rs [hunk 2]
```diff
@@ -33,6 +33,25 @@ pub trait FungibleTokenReceiver {
...
+
+    fn ft_transfer_call(
+        &mut self,
+        receiver_id: ValidAccountId,
```
Comment (evgenykuzyakov) [inline near-contract-standards/src/fungible_token/core_impl.rs:42]:
```suggestion
        receiver_id: AccountId,
```

File: near-contract-standards/src/fungible_token/core_impl.rs [hunk 3]
```diff
@@ -33,6 +33,25 @@ pub trait FungibleTokenReceiver {
...
+        amount: U128,
+        memo: Option<String>,
+        msg: String,
+    ) -> PromiseOrValue<U128>;
+
+    /// Returns the total supply of the token in a decimal string representation.
+    fn ft_total_supply(&self) -> U128;
+
+    /// Returns the balance of the account. If the account doesn't exist must returns `"0"`.
+    fn ft_balance_of(&self, account_id: ValidAccountId) -> U128;
```
Comment (evgenykuzyakov) [inline near-contract-standards/src/fungible_token/core_impl.rs:52]:
```suggestion
    fn ft_balance_of(&self, account_id: AccountId) -> U128;
```

**Review by evgenykuzyakov — CHANGES_REQUESTED** [review]

**Review by evgenykuzyakov — APPROVED** [review]

**Pull request merged**